File: android-katha-ai/app/src/main/java/com/rork/kathaai/ui/screens/CreditPackSheet.kt
package com.rork.kathaai.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Close
import androidx.compose.material.icons.outlined.CreditCard
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.rork.kathaai.model.CreditPack
import com.rork.kathaai.ui.components.PrimaryCTA
import com.rork.kathaai.ui.theme.KathaTheme
import com.rork.kathaai.viewmodel.AppViewModel
import com.rork.kathaai.viewmodel.KathaUiState

@Composable
fun CreditPackSheet(
    state: KathaUiState,
    viewModel: AppViewModel,
    preselectedId: String?,
    onClose: () -> Unit
) {
    var selectedId by remember { mutableStateOf(preselectedId ?: CreditPack.value.id) }
    val selectedPack = CreditPack.all.firstOrNull { it.id == selectedId } ?: CreditPack.value

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(topStart = 24.dp, topEnd = 24.dp))
            .background(KathaTheme.surface)
            .padding(bottom = 32.dp)
    ) {
        // Drag handle
        Box(
            modifier = Modifier
                .padding(top = 12.dp)
                .size(width = 40.dp, height = 4.dp)
                .clip(RoundedCornerShape(2.dp))
                .background(KathaTheme.border)
                .align(Alignment.CenterHorizontally)
        )

        // Header
        Row(
            modifier = Modifier.fillMaxWidth().padding(horizontal = 20.dp, vertical = 16.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Column(modifier = Modifier.weight(1f)) {
                Text("Buy credits", fontSize = 24.sp, fontWeight = FontWeight.Bold, color = KathaTheme.textPrimary)
                Text("One-time purchase. Credits never expire.", fontSize = 13.sp, color = KathaTheme.textSecondary)
            }
            Box(
                modifier = Modifier.size(32.dp).clip(CircleShape).background(KathaTheme.border.copy(alpha = 0.3f)).clickable { onClose() },
                contentAlignment = Alignment.Center
            ) {
                Icon(Icons.Outlined.Close, "Close", tint = KathaTheme.textSecondary, modifier = Modifier.size(14.dp))
            }
        }

        // Pack rows
        Column(modifier = Modifier.fillMaxWidth().padding(horizontal = 20.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
            CreditPack.all.forEach { pack ->
                PackRow(pack = pack, isSelected = pack.id == selectedId) {
                    selectedId = pack.id
                }
            }
        }

        Spacer(Modifier.height(16.dp))

        // Bottom CTA
        Column(modifier = Modifier.fillMaxWidth().padding(horizontal = 20.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
            val ctaLabel = if (state.isPurchasingPack) "Processing…" else "Buy ${selectedPack.name} · ${selectedPack.price}"
            PrimaryCTA(title = ctaLabel) {
                viewModel.purchaseCreditPack(selectedPack)
            }
            Text("Payment charged to your Google account.", fontSize = 11.sp, color = KathaTheme.textTertiary, textAlign = TextAlign.Center, modifier = Modifier.fillMaxWidth())
        }
    }
}

@Composable
private fun PackRow(pack: CreditPack, isSelected: Boolean, onClick: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(12.dp))
            .background(if (isSelected) KathaTheme.accentSoft.copy(alpha = 0.3f) else KathaTheme.surface)
            .border(if (isSelected) 1.5.dp else 1.dp, if (isSelected) KathaTheme.accent else KathaTheme.border, RoundedCornerShape(12.dp))
            .clickable { onClick() }
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Box(modifier = Modifier.size(40.dp), contentAlignment = Alignment.Center) {
            repeat(minOf(pack.coinCount, 3)) { i ->
                Icon(
                    androidx.compose.material.icons.Icons.Outlined.CreditCard, null,
                    tint = KathaTheme.accent,
                    modifier = Modifier.size((16 - i * 2).dp)
                )
            }
        }
        Spacer(Modifier.width(12.dp))
        Column(modifier = Modifier.weight(1f)) {
            Text(pack.name, fontSize = 15.sp, fontWeight = FontWeight.SemiBold, color = KathaTheme.textPrimary)
            Text("${pack.credits} credits", fontSize = 12.sp, color = KathaTheme.textSecondary)
        }
        Text(pack.price, fontSize = 15.sp, fontWeight = FontWeight.Bold, color = KathaTheme.textPrimary)
        Spacer(Modifier.width(12.dp))
        Box(
            modifier = Modifier.size(22.dp).clip(CircleShape).border(2.dp, if (isSelected) KathaTheme.accent else KathaTheme.border, CircleShape),
            contentAlignment = Alignment.Center
        ) {
            if (isSelected) {
                Box(modifier = Modifier.size(12.dp).clip(CircleShape).background(KathaTheme.accent))
            }
        }
    }
}

File: ios-katha-ai-create-stories/KathaAICreateStories/Views/CreditPackSheet.swift
//
//  CreditPackSheet.swift
//  KathaAICreateStories
//

import SwiftUI

// MARK: - Credit Pack Purchase Sheet

struct CreditPackSheet: View {
    @Environment(AppState.self) private var appState
    @State private var selectedPackId: String

    init(preselectedId: String?) {
        _selectedPackId = State(initialValue: preselectedId ?? CreditPack.value.id)
    }

    private var selectedPack: CreditPack {
        CreditPack.all.first { $0.id == selectedPackId } ?? CreditPack.value
    }

    var body: some View {
        VStack(spacing: 0) {
            // Drag handle
            Capsule()
                .fill(KathaTheme.border)
                .frame(width: 40, height: 4)
                .padding(.top, KathaTheme.Spacing.m)
                .padding(.bottom, KathaTheme.Spacing.s)

            // Header
            HStack {
                VStack(alignment: .leading, spacing: 4) {
                    Text("Buy credits")
                        .font(.system(size: 24, weight: .bold))
                        .foregroundStyle(KathaTheme.textPrimary)
                    Text("One-time purchase. Credits never expire.")
                        .font(.system(size: 13))
                        .foregroundStyle(KathaTheme.textSecondary)
                }
                Spacer()
                Button {
                    Haptics.light()
                    appState.closeCreditPackSheet()
                } label: {
                    Image(systemName: "xmark")
                        .font(.system(size: 14, weight: .semibold))
                        .foregroundStyle(KathaTheme.textSecondary)
                        .frame(width: 32, height: 32)
                        .background(Circle().fill(KathaTheme.border.opacity(0.3)))
                }
            }
            .padding(.horizontal, KathaTheme.Spacing.l)
            .padding(.top, KathaTheme.Spacing.m)

            // Pack tiles
            VStack(spacing: KathaTheme.Spacing.s) {
                ForEach(CreditPack.all) { pack in
                    packRow(pack)
                }
            }
            .padding(.horizontal, KathaTheme.Spacing.l)
            .padding(.top, KathaTheme.Spacing.l)

            Spacer()

            // Bottom pinned CTA
            VStack(spacing: KathaTheme.Spacing.s) {
                PrimaryCTA(
                    title: appState.isPurchasingPack
                        ? "Processing…"
                        : "Buy \(selectedPack.name) · \(selectedPack.price)"
                ) {
                    Task { await appState.purchaseCreditPack(selectedPack) }
                }
                .disabled(appState.isPurchasingPack)

                Text("Payment charged to your Apple ID.")
                    .font(.system(size: 11))
                    .foregroundStyle(KathaTheme.textTertiary)
            }
            .padding(.horizontal, KathaTheme.Spacing.l)
            .padding(.bottom, KathaTheme.Spacing.l)
        }
        .background(KathaTheme.surface)
        .clipShape(RoundedRectangle(cornerRadius: KathaTheme.Radius.xxl))
    }

    private func packRow(_ pack: CreditPack) -> some View {
        let isSelected = pack.id == selectedPackId

        return Button {
            Haptics.light()
            withAnimation(.easeInOut(duration: 0.2)) {
                selectedPackId = pack.id
            }
        } label: {
            HStack(spacing: KathaTheme.Spacing.m) {
                ZStack {
                    ForEach(0..<min(pack.coinCount, 5), id: \.self) { i in
                        Image(systemName: "dollarsign.circle.fill")
                            .font(.system(size: 16 - CGFloat(i) * 2))
                            .foregroundStyle(KathaTheme.accent)
                            .offset(x: CGFloat(i) * 5, y: CGFloat(i) * 3)
                    }
                }
                .frame(width: 40, height: 32)

                VStack(alignment: .leading, spacing: 2) {
                    Text(pack.name)
                        .font(.system(size: 15, weight: .semibold))
                        .foregroundStyle(KathaTheme.textPrimary)
                    Text("\(pack.credits) credits")
                        .font(.system(size: 12))
                        .foregroundStyle(KathaTheme.textSecondary)
                }

                Spacer()

                Text(pack.price)
                    .font(.system(size: 15, weight: .bold))
                    .foregroundStyle(KathaTheme.textPrimary)

                ZStack {
                    Circle()
                        .stroke(isSelected ? KathaTheme.accent : KathaTheme.border, lineWidth: 2)
                        .frame(width: 22, height: 22)
                    if isSelected {
                        Circle()
                            .fill(KathaTheme.accent)
                            .frame(width: 12, height: 12)
                    }
                }
            }
            .padding(.horizontal, KathaTheme.Spacing.l)
            .padding(.vertical, KathaTheme.Spacing.m)
            .background(
                RoundedRectangle(cornerRadius: KathaTheme.Radius.m)
                    .fill(isSelected ? KathaTheme.accentSoft.opacity(0.3) : KathaTheme.surface)
                    .overlay(
                        RoundedRectangle(cornerRadius: KathaTheme.Radius.m)
                            .stroke(isSelected ? KathaTheme.accent : KathaTheme.border, lineWidth: isSelected ? 1.5 : 1)
                    )
            )
        }
        .buttonStyle(PressScaleStyle(scale: 0.98))
    }
}
